### ISSUE TITLE
dimulski - A malicious user can lock the prize for winning a raffle provided by the admin when a raffle is canceled

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-08-winnables-raffles-judging#147]
dimulski

High

# A malicious user can lock the prize for winning a raffle provided by the admin when a raffle is canceled

### Summary

An admin of the Winnables protocol can create a raffle by first calling the [lockNFT()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L148-L165), [lockETH()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L172-L188), or the [lockTokens()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L196-L213) functions in order to lock the prize that the winner of the raffle will receive. Then a CCIP message will be sent to the ``WinnablesTicketManager.sol`` contract, and set the **RaffleStatus** for the raffle id to **PRIZE_LOCKED**. Then the admin has to call the [createRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L252C5-L274) function in order to fully set the raffle, and users can start buying tickets. If the raffle has ended and the bought tickets are below the **minTicketsThreshold** set by the admin, the raffle can be canceled by everyone. However the [cancelRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L278-L288) function takes an arbitrary **prizeManager** and **chainSelector** as arguments:
```solidity
    function cancelRaffle(address prizeManager, uint64 chainSelector, uint256 raffleId) external {
        _checkShouldCancel(raffleId);

        _raffles[raffleId].status = RaffleStatus.CANCELED;
        _sendCCIPMessage(
            prizeManager,
            chainSelector,
            abi.encodePacked(uint8(CCIPMessageType.RAFFLE_CANCELED), raffleId)
        );
        IWinnablesTicket(TICKETS_CONTRACT).refreshMetadata(raffleId);
    }
```
As can be seen from the above code snippet the [cancelRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L278-L288) function also sets the **_raffles[raffleId].status** to **RaffleStatus.CANCELED**. As can be seen from the [_checkShouldCancel()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L434-L441) function, if the raffleStatus is **CANCALED**, the function will revert. 
```solidity
    function _checkShouldCancel(uint256 raffleId) internal view {
        Raffle storage raffle = _raffles[raffleId];
        if (raffle.status == RaffleStatus.PRIZE_LOCKED) return;
        if (raffle.status != RaffleStatus.IDLE) revert InvalidRaffle();
        if (raffle.endsAt > block.timestamp) revert RaffleIsStillOpen();
        uint256 supply = IWinnablesTicket(TICKETS_CONTRACT).supplyOf(raffleId);
        if (supply > raffle.minTicketsThreshold) revert TargetTicketsReached();
    }
```
A malicious user can call the [cancelRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L278-L288) function in order to cancel a raffle, and supply a **prizeManager** address that is different than the address of the ``WinnablesPrizeManager.sol`` deployed contract where the admin initially locked the reward for the winner of the raffle. A malicious actor can simply deploy another contract that can receive CCIP messages (so the [cancelRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L278-L288) function doesn't revert). The  [cancelRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L278-L288) function can't be called again for the same raffle ID, and the funds that the admin initially locked in the ``WinnablesPrizeManager.sol`` contract will be locked forever, resulting in a loss of funds for the admin. 

### Root Cause

The [cancelRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L278-L288) function can be called by anyone, with arbitrary **prizeManager** and **chainSelector** arguments.

### Internal pre-conditions

1. Raffle with ID 1 has expired and the bought tickets are below the **minTicketsThreshold** set by the admin, now the raffle can be canceled.

### Attack Path

1. Raffle with ID 1 has expired and the bought tickets are below the **minTicketsThreshold** set by the admin, now the raffle can be canceled.
2. A malicious user deploys a contract which can receive CCIP messages, lets call it **ContractB**
3. The malicious user calls the [cancelRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L278-L288) function but instead of providing the correct address of the ``WinnablesPrizeManager.sol`` contract, where the admin initially locked funds, he provides the address of **ContractB**. 

### Impact

The funds that the admin locked as a reward for the winner of the raffle, either an NFT, ETH, or some ERC20 token will be locked forever in the ``WinnablesPrizeManager.sol`` contract.

### Mitigation

Consider setting up the **prizeManager** and **chainSelector** in the constructor, or create a function that can be called only once  by the admin with the correct parameters.  Use those parameters in the [cancelRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L278-L288) function, don't allow users to provide arbitrary parameters. 

Duplicate of #50

**Comment by AtanasDimulski** [comment]
Escalate,
This issue is not a duplicate of #50, the root cause and attack are completely different than the ones mentioned in #50 . This report describes an issue where if a raffle is canceled by a malicious user, the funds provided by the admin as a reward will be locked forever.

**Comment by sherlock-admin3** [comment]
> Escalate,
> This issue is not a duplicate of #50, the root cause and attack are completely different than the ones mentioned in #50 . This report describes an issue where if a raffle is canceled by a malicious user, the funds provided by the admin as a reward will be locked forever.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by matejdrazic** [comment]
Please see my issue #158 because its been wrongly duped with #50. My issue is about lack of access control on the cancel function.

**Comment by mystery0x** [comment]
> Escalate, This issue is not a duplicate of #50, the root cause and attack are completely different than the ones mentioned in #50 . This report describes an issue where if a raffle is canceled by a malicious user, the funds provided by the admin as a reward will be locked forever.

Still, the similar mitigation as in #50 when implemented will have it circumvented.

**Comment by WangSecurity** [comment]
Indeed issue #50 and this report happen in different functions. But, both identify almost the same scenario where you call a function `propagateRaffleWinner` in #50 and `cancelRaffle` in this report with the incorrect prize manager (by incorrect I mean not the one that should be used) causing the lock of winner's prize. Is it correct or am I missing a larger difference than the issue happens in 2 different functions?

**Comment by AtanasDimulski** [comment]
Hey @WangSecurity, both issues describe the same attack vector. However, in this one the impact is different - the admin will lose the funds he provided as a potential reward. The function is different, and the mitigation provided in #50 won't mitigate the issue described in this report. Although it is the same attack vector, I do believe these issues are not duplicates. If you consider them as such keep in mind you have two options: there are reports that describe only 1 of them, missing the other is still a critical vulnerability so the reports that mention only one should be given partial-50, or both issues should be considered as separate ones. I believe the second option is the correct one. Thanks for your time!

**Comment by WangSecurity** [comment]
Thank you for the elaboration, to clarify some last points to make sure I don't miss anything. The root cause of the issues is the same: not validating the prizeManager address and chain selector.
About the fixes, both issues can be mitigated with the same fix of validating the prizeManager address and chain selector, no? Of course, they have to fixed separately because they happen in different functions, but the fix itself can be the same. 

Am I missing anything here?

**Comment by AtanasDimulski** [comment]
Hey @WangSecurity sorry for the late reply missed this one. If you consider the same attack vector to be the same root, then yes both issues have the same root - not validating the prizeManager address and chain selector and both functions are permissionless. Theoretically, the fix is pretty much the same, although it has to be implemented for both functions. I just want to reiterate the impact is different, and quote this from the readme: *The exception to this would be if underlying code implementations OR impact OR the fixes are different*. Also reports that identify only one function shouldn't receive full credit, because if the protocol fixes only one of the functions the vulnerability still exists, I don't know how you can grade those reports and be fair to all wardens. Because for example #50 only describes a problem in the propagateRaffleWinner() function, and how to mitigate the issue in that function. So if only #50 was submitted, and my issue and other issues that report a problem with the cancelRaffle() function were not submited, and the protocol fixed only the propagateRaffleWinner() function the protocol would still be vulnerable. Thanks for your time!

**Comment by WangSecurity** [comment]
That's a fair argument, but given the fact the root cause of both issues and the fixes are the same (although, have to be fixed separately), they have to remain duplicates. I see that indeed the rules say even if the impact is different, they may be treated separately, but it doesn't mean they should be. Hence, my decision remains that they have to remain duplicates based on the root cause and the fix being the same. Planning to reject the escalation and leave the issue as it is.

**Comment by WangSecurity** [comment]
Result:
High
Duplicate of #50

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [AtanasDimulski](https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/147/#issuecomment-2329447411): rejected
